File: leetcode/exercises/2.2.4.medium.generate_parenthesis_1.go
// -----------------------
// Created     : 26/11/2025
// Last Edited : 26/11/2025 
// Topics      : 
// Big O       :
// Problem Id  : 150
// References  : https://www.youtube.com/watch?v=s9fokUqJ76A
// Notes       : Man this backtracking stuff is tuff
// -----------------------

package main

import (
    "fmt"
)

func generateParenthesis(n int) []string {

    res := []string{}
    stack := []rune{}

    var backtracking func(open, closed int)
    backtracking = func(open int, closed int) {

        if open == n && closed == n {
            current := ""
            for _, v := range stack {
                current += string(v)
            }
            res = append(res, current)
            return
        }

        if open < n {
            stack = append(stack, '(')
            backtracking(open + 1, closed)
            stack = stack[:len(stack)-1]
        }

        if closed < open {
            stack = append(stack, ')')
            backtracking(open, closed+1)
            stack = stack[:len(stack)-1]
        }

    }

    backtracking(0, 0)
    return res
}


func main() {
    fmt.Println(generateParenthesis(3))
}

File: leetcode/exercises/2.2.4.medium.generate_parenthesis_1.js
// -----------------------
// Created     : 26/11/2025
// Last Edited : 26/11/2025 
// Topics      : 
// Big O       :
// Problem Id  : 150
// References  :
// Notes       : Man this backtracking stuff is tuff
// -----------------------

/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function(n) {
    const stack = [];
    const res = [];
    var backtracking = function(open, close) {
        if (open === n && close === n) {
            res.push(stack.join(""));
            return
        }

        if (open < n) {
            stack.push("(");
            backtracking(open + 1, close);
            stack.pop();
        }

        if (close < open) {
            stack.push(")");
            backtracking(open, close + 1);
            stack.pop();
        }

    }
    backtracking(0, 0);
    return res
}

console.log(generateParenthesis(3));
